core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1158,9 +1158,14 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
     @throws(classOf[Exception])
     def handleNewSession() {
       info("ZK expired; shut down all controller components and try to re-elect")
-      onControllerResignation()
-      inLock(controllerContext.controllerLock) {
-        controllerElector.elect
+      if (controllerElector.getControllerID() != config.brokerId) {
+        onControllerResignation()
+        inLock(controllerContext.controllerLock) {
+          controllerElector.elect
+        }
+      } else {
```

Comment:
@pengwei-li : That's the part that I still don't understand. If the ephemeral node is created by LeaderChangeListener.handleDataDeleted on the current session, ZKCheckedEphemeral.create() should succeed in KafkaController.handleNewSession (through controllerElector.elect). Could you figure out why ZKNodeExistException is thrown instead? That will help us determine where the fix should be.